kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,231 @@
+        } catch (Exception e) {
+            logger.error("Search KieServerState with id [{}] failed.", id);
+        }
+        return hasIt;
+    }
+
+    public KieServerState delete(String id) {
+        KieServerState state = load(id);
+
+        try (OpenShiftClient client = createOpenShiftClient()) {
+            if (client.deploymentConfigs().withName(id).get() == null) {
+                client.configMaps().withName(id).delete();
+            } else {
+                /**
+                 * If an associated DeploymentConfig exists, 
+                 * ignore the delete request for ConfigMap. 
+                 */
+                logger.error("Can not delete attached KieServerState with id [{}].", id);
+            }
+        } catch (Exception e) {
+            logger.error("Delete KieServerState with id [{}] failed.", id);
+        }
+        return state;
+    }
+
+    @Override
+    public boolean isKieServerReady() {
+        return KieServerLocator.getInstance().isKieServerReady();
+    }
+
+    protected String retrieveKieServerId(KieServerState kieServerState) {
+        String kssServerId = null;
+
+        try {
+            kssServerId = kieServerState.getConfiguration().getConfigItemValue(KieServerConstants.KIE_SERVER_ID);
+        } catch (Exception e) {
+            logger.error("Failed to retrieve server id from KieServerState", e);
+        }
+
+        if (kssServerId == null || kssServerId.length() == 0) {
+            throw new IllegalArgumentException("Invalid KieServerId: Can not be null or empty.");
+        }
+        return kssServerId;
+    }
+
+    protected void createOrUpdate(String serverId, KieServerState kieServerState, boolean isUpdate) {
+        try (KubernetesClient client = createKubernetesClient()) {
+            String stateXML = xs.toXML(kieServerState);
+
+            ConfigMap cm = client.configMaps().withName(serverId).get();
+
+            if (cm == null) {
+                if (isUpdate) {
+                    throw new IllegalStateException("KieServerState ConfigMap must exist before update.");
+                } else {
+                    cm = client.configMaps().create(new ConfigMapBuilder()
+                                                                          .withNewMetadata()
+                                                                          .withName(serverId)
+                                                                          .endMetadata()
+                                                                          .withData(Collections.singletonMap(CFG_MAP_DATA_KEY, stateXML))
+                                                                          .build());
+                }
+            } else {
+                cm.setData(Collections.singletonMap(CFG_MAP_DATA_KEY, stateXML));
+            }
+
+            if (!isUpdate && cm != null) {
+                logger.info("Create new ConfigMap action ignored. ConfigMap for KieServer [{}] exists.", serverId);
```

Comment:
What should happen in case you try to create a config map when it already exists?
From conditions above it should set new data regardless of if this is update or create (see line 205). Also here the cm object can never be null.
I would recommend to split create and update methods to make the code easier to read and understand what happens when.

Reply:
Refactored method into subclass
https://github.com/rhtevan/droolsjbpm-integration/blob/kie-server-ocp-rearch-1/kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java